pkg/spoke/auth/cache_store.go

```diff
@@ -0,0 +1,183 @@
+package auth
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"k8s.io/klog/v2"
+)
+
+func NewExecutorCache() *ExecutorCaches {
+	return &ExecutorCaches{
+		lock:  sync.RWMutex{},
+		items: make(map[string]*DimensionCaches),
+	}
+}
+
+type ExecutorCaches struct {
+	lock sync.RWMutex
+
+	// map key: executor in format of {namespace}/{name}
+	items map[string]*DimensionCaches
+}
+
+type DimensionCaches struct {
+	lock sync.RWMutex
+
+	// map key: hash(Dimension)
+	items map[string]CacheValue
+}
+
+type CacheValue struct {
+	Dimension Dimension
+	Allowed   bool
+}
+
+type Dimension struct {
+	Group         string
+	Version       string
+	Resource      string
+	Namespace     string
+	Name          string
+	ExecuteAction ExecuteAction
+}
+
+// AddNewDimensionCaches adds a dimension caches by the executor if it does not exist, updates it if it exists
+func (c *ExecutorCaches) AddNewDimensionCaches(executor string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	dimensionCaches := &DimensionCaches{
+		lock:  sync.RWMutex{},
+		items: make(map[string]CacheValue),
+	}
+	c.items[executor] = dimensionCaches
+}
+
+func (c *ExecutorCaches) GetDimensionCaches(executor string) (*DimensionCaches, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	item, ok := c.items[executor]
+	return item, ok
+}
+
+func (c *ExecutorCaches) RemoveDimensionCaches(executor string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	delete(c.items, executor)
+}
+
+func (c *ExecutorCaches) GetCacheItems() map[string]*DimensionCaches {
+	c.lock.Lock()
```

Comment:
hrm, why we need this? can  a func get this and mutate items out of lock?
or you want a copy of the items to be returned?

Reply:
I want a copy of the items to be returned. so in the cache controller we can use a for-range loop to update these caches.

Reply:
This has been fixed